idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,1160 @@
+
+                 # recursion
+                 dfs_sum_revenue(set_idx + 1, indices, bidding_indices)
+
+                 # backtrack
+                 indices.pop()
+                 if model_sets[set_ls[set_idx]] in self.bidding_set_names:
+                     bidding_indices.pop()
+
+        for cost in cost_ls:
+            dfs_sum_costs(cost = cost, set_idx = 0, indices = [])
+        dfs_sum_revenue(set_idx = 0, indices = [], bidding_indices = [])
+
+    def compute_bids(self, price_forecasts, date, hour):
+
+        '''
+        Solve the model to bid into the markets. After solving, record the bids
+        from the solve.
+
+        Arguments:
+            price_forecasts: price forecasts needed to solve the bidding problem.
+                            {gen: {LMP scenario: [forecast timeseries] }}
+            date: current simulation date
+            hour: current simulation hour
+
+        Returns:
+            None
+        '''
+
+        # update the price forecasts
+        self._pass_price_forecasts(price_forecasts)
+        self.solver.solve(self.model,tee=True)
+        bids = self._assemble_bids()
+        self.record_bids(bids, date = date, hour = hour)
+
+        ## TODO: when to update bidding model??
+
+        return bids
+
+    def _pass_price_forecasts(self, price_forecasts):
+
+        '''
+        Pass the price forecasts into model parameters.
+
+        Arguments:
+            price_forecasts: price forecasts needed to solve the bidding problem.
+                             {generator: {LMP scenario: [forecast timeseries] }}
+
+        Returns:
+            None
+        '''
+
+        model_sets = self.bidding_model_object.indices
+        energy_price = self.model.energy_price
+
+        aux_dict = {'Generators': None, 'Time': None, 'LMP Scenarios': None}
+
+        for idx1 in self.bidding_set[0]:
+            for idx2 in self.bidding_set[1]:
+                for idx3 in self.bidding_set[2]:
+
+                    aux_dict[model_sets[self.bidding_set[0]]] = idx1
+                    aux_dict[model_sets[self.bidding_set[1]]] = idx2
+                    aux_dict[model_sets[self.bidding_set[2]]] = idx3
+
+                    energy_price[idx1,idx2,idx3] = price_forecasts[aux_dict['Generators']][aux_dict['LMP Scenarios']][aux_dict['Time']]
+
+        return
+
+    def _assemble_bids(self):
+
+        '''
+        This methods extract the bids out of the stochastic programming model and
+        organize them.
+
+        Arguments:
+
+        Returns:
+            bids: the bid we computed. It is a dictionary that has this structure
+            {t: {gen:{power: cost}}}.
+        '''
+
+        model_sets = self.bidding_model_object.indices
+        model_sets_inverse = {}
+        for s, n in model_sets.items():
+            model_sets_inverse[n] = s
+
+        # unpack things
+        model_sets = self.bidding_model_object.indices
+        set_ls = list(model_sets.keys())
```

Comment:
Would it make sense to have an `unpack` method that returned `set_ls`, `power_output`, and `energy_price`?